### PR TITLE
Merge common APIs for Dictionary

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
## Description
Motivation:
Currently the APIs for Dictionary is split in 3 places: `Dictionary`, `BaseImmutableDictionary`, `BaseMutableDictionary`. In order to use them, we need to cast the dictionary first, which is hard to manage and can potentially cause casting error.
E.g. #6174 is caused by casting an immutable dictionary to `BaseMutableDictionary`.
We should move the common read APIs to the root `Dictionary` interface to avoid the casting, and let all types of dictionary support these APIs.

Merge the following common APIs from `BaseImmutableDictionary` and `BaseMutableDictionary` to `Dictionary`:
- `insertionIndexOf`
- `getDictIdsInRange`
- `compare`
- `getMinVal`
- `getMaxVal`
- `getSortedValues`

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/BaseImmutableDictionary.java [hunk 1]
```diff
@@ -81,6 +80,31 @@ public int indexOf(String stringValue) {
...
+  }
+
+  @Override
+  public Comparable getMaxVal() {
+    return (Comparable) get(_length - 1);
+  }
+
+  @Override
+  public Object getSortedValues() {
+    throw new UnsupportedOperationException();
```
Comment (yupeng9) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/BaseImmutableDictionary.java:105]: why is this unsupported? since it returns true for `isSorted `

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/BaseImmutableDictionary.java [hunk 2]
```diff
@@ -81,6 +80,31 @@ public int indexOf(String stringValue) {
     return (index >= 0) ? index : NULL_VALUE_INDEX;
   }
 
+  @Override
+  public IntSet getDictIdsInRange(String lower, String upper, boolean includeLower, boolean includeUpper) {
+    throw new UnsupportedOperationException();
```
Comment (yupeng9) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/BaseImmutableDictionary.java:85]: why can this be supported?

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/BaseImmutableDictionary.java (hunk 1 above)
Comment (yupeng9) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/BaseImmutableDictionary.java:105]: If you want a default impl, then perhaps just remove it, so the subclass must implement it.

Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/BaseImmutableDictionary.java:105]: Added the note: This method is for the stats collection phase when sealing the consuming segment, so not required for regular immutable dictionary within the immutable segment.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java [hunk 3]
```diff
@@ -40,14 +42,52 @@
...
 
+  /**
+   * Returns the insertion index of the string representation of the value in the dictionary. This method follows the
+   * same behavior as in {@link Arrays#binarySearch(Object[], Object)}. All sorted dictionary should support this
+   * method. This method is for the range predicate evaluation.
+   */
+  int insertionIndexOf(String stringValue);
+
+  /**
+   * Returns a set of dictIds in the given value range, where lower/upper bound can be "*" which indicates unbounded
```
Comment (chenboat) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java:64]: is the range inclusive or exclusive the in lower and upper bounds?

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java [hunk 4]
```diff
@@ -40,14 +42,52 @@
...
   /**
    * Returns the index of the string representation of the value in the dictionary, or {@link #NULL_VALUE_INDEX} (-1) if
-   * the value does not exist. This API is for cross-type predicate evaluation.
+   * the value does not exist. This method is for the cross-type predicate evaluation.
    */
   int indexOf(String stringValue);
 
+  /**
+   * Returns the insertion index of the string representation of the value in the dictionary. This method follows the
+   * same behavior as in {@link Arrays#binarySearch(Object[], Object)}. All sorted dictionary should support this
```
Comment (chenboat) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java:58]: NIT: dictionary-> dictionaries

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java [hunk 5]
```diff
@@ -40,14 +42,52 @@
...
+  int insertionIndexOf(String stringValue);
+
+  /**
+   * Returns a set of dictIds in the given value range, where lower/upper bound can be "*" which indicates unbounded
+   * range. All unsorted dictionary should support this method. This method is for the range predicate evaluation.
+   */
+  IntSet getDictIdsInRange(String lower, String upper, boolean includeLower, boolean includeUpper);
+
+  /**
+   * Returns the comparison result of value for dictId 1 and dictId 2, i.e. {@code value1.compareTo(value2)}.
```
Comment (chenboat) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java:70]: is it string value comparison?

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/BaseImmutableDictionary.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/BaseImmutableDictionary.java:85]: This method is for the unsorted dictionary. Added the note

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java [hunk 6]
```diff
@@ -40,14 +42,52 @@
...
+   */
+  int compare(int dictId1, int dictId2);
+
+  /**
+   * Returns the minimum value in the dictionary. Note that for type BYTES, {@code ByteArray} will be returned.
```
Comment (chenboat) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java:75]: what happens if the dictionary is empty?

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java (hunk 4 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java:58]: Done

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java (hunk 5 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java:70]: No, actual value. Updated the javadoc

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java (hunk 6 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java:75]: Good question. It is undefined, and should not be called when the dictionary is empty. Updated the javadoc.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java [hunk 7]
```diff
@@ -56,12 +56,12 @@ protected FilterBlock getNextBlock() {
 
     int firstRangeId;
     int lastRangeId;
-    if (_rangePredicateEvaluator instanceof OfflineDictionaryBasedRangePredicateEvaluator) {
+    if (_rangePredicateEvaluator instanceof SortedDictionaryBasedRangePredicateEvaluator) {
```
Comment (chenboat) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java:59]: is this change related to this API refactoring?

Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java:59]: Yes. We should determine the evaluator by whether the dictionary is sorted, instead of the instance of the dictionary.

Comment (chenboat) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java:59]: I mean I did not see any change to OfflineDictionaryBasedRangePredicateEvaluator nor SortedDictionaryBasedRangePredicateEvaluator in this PR. Is this an existing bug for range predicate?

**Review by yupeng9 — APPROVED** [review]

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java (hunk 7 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java:59]: Oh, that is under the `RangeIndexBasedFilterOperator`

Comment (chenboat) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java:59]: Ah. Got it. You made changes in RangePredicateEvaluatorFactory.java.

**Review by chenboat — APPROVED** [review]

**Comment by deemoliu** [comment]
Hello @Jackie-Jiang Can we include this fix in [Apache Pinot (incubating) 0.6.0](https://github.com/apache/incubator-pinot/releases/tag/release-0.6.0-rc0)? 

cc'ed @yupeng9 @chenboat

**Comment by Jackie-Jiang** [comment]
@deemoliu I think we will cut the release on 10/30. @jackjlli can you please confirm?

**Comment by jackjlli** [comment]
Yes, the new release 0.6.0 will be cut on 10/30.
